### ISSUE TITLE
Bot couldn't sell because of insufficient fund

### ISSUE THREAD

**Issue by pan-long** [freqtrade/freqtrade#1371]
I was synced to HEAD yesterday (but I had the same issue some time ago).

I had an order that the bot couldn't sell because of insufficient fund. I think it's because either the bot did wrong calculation of the amount or the exchange is doing some funny stuff.

Logs when the order was fulfilled:
```
2018-11-29 03:05:08,831 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=xx, pair=BCHSV/USDT, amount=xxx, open_rate=99.10000000, open_since=just now)
2018-11-29 03:05:09,740 - freqtrade.freqtradebot - WARNING - amount xxx does not match amount xxx
2018-11-29 03:05:09,740 - freqtrade.freqtradebot - WARNING - could not update trade amount: Half bought? Amounts don't match
2018-11-29 03:05:09,740 - freqtrade.persistence - INFO - Updating trade (id=xx) ...
2018-11-29 03:05:09,741 - freqtrade.persistence - INFO - LIMIT_BUY has been fulfilled for Trade(id=xx, pair=BCHSV/USDT, amount=xxx, open_rate=99.10000000, open_since=just now).
```

Logs when the bot was trying to sell:
```
2018-11-29 03:09:29,427 - freqtrade.freqtradebot - INFO - checking sell
2018-11-29 03:09:29,910 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market BCHSV/USDT.Tried to sell amount xx at rate 100.02 (total xx).Message: binance {"code":-2010,"msg":"Account has insufficient balance for requested action."}
```

The code that is throwing the error, any reason that instead of using the latest amount getting from exchange, we still choose to use the one saved when buying?
https://github.com/freqtrade/freqtrade/blob/200484ab8ba85308e6e759bbec613dcba0e83d37/freqtrade/freqtradebot.py#L603

**Comment by xmatthias** [comment]
i think the critical logs are in the buy section:

```
2018-11-29 03:05:09,740 - freqtrade.freqtradebot - WARNING - amount xxx does not match amount xxx
2018-11-29 03:05:09,740 - freqtrade.freqtradebot - WARNING - could not update trade amount: Half bought? Amounts don't match
```

this seems odd - could you please compare the contents of the database fort this trade with the binance UI to check what did change?

The best thing is that you have the whole logs available - this will help a TON in investigating this issue.

I wonder if it's got to do with the rounding / truncating applied (but that's just thinking out loud so i don't forget that train of thought)

**Comment by mishaker** [comment]
@xmatthias while we are on it ... do we really need `symbol_price_prec` and `symbol_amount_prec`? 
They are covered by CCXT: 
```python
def price_to_precision(self, symbol, price):
        return self.decimal_to_precision(price, ROUND, self.markets[symbol]['precision']['price'], self.precisionMode)

def amount_to_precision(self, symbol, amount):
        return self.decimal_to_precision(amount, TRUNCATE, self.markets[symbol]['precision']['amount'], self.precisionMode)
```

**Comment by xmatthias** [comment]
i was honestly never convince that we do, however the original PR had a solid explanation as to why ... but i don't remember that :( 

will try to find the PR for reference - hopefully that'll give us some insight

**Comment by pan-long** [comment]
Unfortunately, I lost the full logs. I just checked my binance orders, the amount showing there is 

xxx.51969

and I remember the amount showing in the logs was something like

xxx.5197xxxx

I don't think it's because of rounding/truncating because I am sure it wasn't xxx.51970000

I have actually done some [local changes](https://github.com/pan-long/freqtrade/commit/5c6f2b2f9c16e6d159587e5091b5d5a7b5d7b56d) and the bot has been running fine for days. But I still don't know how the bot got the wrong amount though.

**Comment by mishaker** [comment]
@xmatthias have tons of order failed because we don't use `self.price_to_precision`and `self. amount_to_precision` of ccxt:
```
Message: binance order price is invalid, i.e. exceeds allowed price precision, exceeds min price or max price limits or is invalid float value in general, use self.price_to_precision(symbol, amount) {"code":-1013,"msg":"Filter failure: PRICE_FILTER"}
```
I read the PR, I am not convinced ... 
If @creslinux does not react, then I think we should remplace those functions with CCXT.

**Comment by creslinux** [comment]
The precision filters are fine, I use them a lot.

I would suggest from the list of filters binance provide FT runs it's order
against the ones not being checked.

Min notional, Min and Max amount and price as example.

Precision will not help an order that's not passing those.

**Comment by creslinux** [comment]
Check order amounts/quantities.

The way edge is calculating risk not on full portfolio worth, it's position
sizes are going to be tiny sums which may do not meet minimum amounts
binance accepts.

Also check the info{} dict inside ccxt message. It will contain the actual
error message which filter is rejecting the order.

**Comment by xmatthias** [comment]
@mishaker as mentioned before, i'm not tied to our logic to calculate precisions and would prefer to relie on ccxt handling that.

and yes - also the other info (min-notational, ...) will absolutely need to be considered. ccxt does have that available easily and unified - in `fetch_markets()` (i think)

**Comment by creslinux** [comment]
Guys, strongly suggest create a binance subclass and be done with it.
Youre going to end up with a hundred IF's to manage all the specific
binance stuff ccxt does not give you.
You need ALL the filters to not bounce trades frequently - drop it in the
subclass.

CCXT is the lowest common denominator - you cant trade on it without being
at such a disadvantage to other traders.
Or your money at such a disadvantage to be more accurate

AS example of some the things you're going to end up having to manage
explicitly for binance that are not common in ccxt.
if id == bianance:
 -- this filter
 -- that filter
 -- another filter
if id == binance
   order-type-a param timinginforce =
   order-type-b param timinginforce =
if id == binance
   stop_loss = stop_loss_limit
   param  stop_price ==
if id == binance
  ulrlib =  i need all the candle data
if id == binance
  my fees == arethis... =
etc etc ect

**Comment by mishaker** [comment]
For the record: recent bombardments of price filter error was due to a sudden change in Binance API which is now handled in the following: https://github.com/ccxt/ccxt/commit/dc3bd40b4c6e24fc23211b1aba29df8a5ac89947

**Comment by creslinux** [comment]
That's not recent, it's Ccxt code being rubbish - I had those in my filters
class couple days into Dec.
Here is the part of the API i emailed FT start of the week, explicitly
saying to check the price and qty in the info{} message returned.

Binance tell the filter is failed on, CCXT hides it.
Symbol filters
<https://github.com/binance-exchange/binance-official-api-docs/blob/master/rest-api.md#price_filter>
PRICE_FILTER

The PRICE_FILTER defines the price rules for a symbol. There are 3 parts:

   - minPrice defines the minimum price/stopPrice allowed; disabled on
   minPrice == 0.
   - maxPrice defines the maximum price/stopPrice allowed; disabled on
   maxPrice == 0.
   - tickSize defines the intervals that a price/stopPrice can be
   increased/decreased by; disabled on tickSize == 0.

**Comment by asheetov** [comment]
The same issue here ```binance {"code":-2010,"msg":"Account has insufficient balance for requested action."}```

It looks like a floating bug. Sometimes selling works fine, sometimes doesn't.
Good one:
```
2018-12-20 04:58:00,073 - freqtrade.rpc.telegram - INFO - Executing handler: _forcebuy for chat_id: 647795322
2018-12-20 04:58:00,591 - freqtrade.freqtradebot - INFO - Using Last Ask / Last Price
2018-12-20 04:58:01,603 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': buy, 'exchange': 'Binance', 'pair': 'BCHABC/USDT', 'market_url': 'https://www.binance.com/tradeDetail.html?symbol=BCHABC_USDT', 'limit': 157.09, 'stake_amount': 100, 'stake_currency': 'USDT', 'fiat_currency': 'USD'}
2018-12-20 04:58:02,117 - freqtrade.wallets - INFO - Wallets synced ...
2018-12-20 04:58:06,268 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=6, pair=BCHABC/USDT, amount=0.63657776, open_rate=157.09000000, open_since=just now)
2018-12-20 04:58:07,283 - freqtrade.freqtradebot - INFO - Applying fee on amount for Trade(id=6, pair=BCHABC/USDT, amount=0.63657776, open_rate=157.09000000, open_since=just now) (from 0.63657 to 0.63593343) from Trades
2018-12-20 04:58:07,284 - freqtrade.persistence - INFO - Updating trade (id=6) ...
2018-12-20 04:58:07,284 - freqtrade.persistence - INFO - LIMIT_BUY has been fulfilled for Trade(id=6, pair=BCHABC/USDT, amount=0.63593343, open_rate=157.09000000, open_since=just now).

2018-12-20 05:16:23,409 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': sell, 'exchange': 'Binance', 'pair': 'BCHABC/USDT', 'gain': 'loss', 'market_url': 'https://www.binance.com/tradeDetail.html?symbol=BCHABC_USDT', 'limit': 154.3, 'amount': 0.63593343, 'open_rate': 157.09, 'current_rate': 154.38, 'profit_amount': -1.8723788, 'profit_percent': -0.01874276, 'sell_reason': 'trailing_stop_loss', 'stake_currency': 'USDT', 'fiat_currency': 'USD'}
2018-12-20 05:16:24,093 - freqtrade.freqtradebot - INFO - executed sell, reason: SellType.TRAILING_STOP_LOSS
2018-12-20 05:16:24,215 - freqtrade.wallets - INFO - Wallets synced ...
2018-12-20 05:16:25,224 - freqtrade.wallets - INFO - Wallets synced ...
2018-12-20 05:16:26,891 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=6, pair=BCHABC/USDT, amount=0.63593343, open_rate=157.09000000, open_since=18 minutes ago)
2018-12-20 05:16:27,402 - freqtrade.persistence - INFO - Updating trade (id=6) ...
2018-12-20 05:16:27,402 - freqtrade.persistence - INFO - Marking Trade(id=6, pair=BCHABC/USDT, amount=0.63593343, open_rate=157.09000000, open_since=closed) as closed as the trade is fulfilled and found no open orders for it.
```

Bad one:
```
2018-12-20 05:38:36,028 - freqtrade.rpc.telegram - INFO - Executing handler: _forcebuy for chat_id: 647795322
2018-12-20 05:38:36,156 - freqtrade.freqtradebot - INFO - Using Last Ask / Last Price
2018-12-20 05:38:37,223 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': buy, 'exchange': 'Binance', 'pair': 'BCHABC/USDT', 'market_url': 'https://www.binance.com/tradeDetail.html?symbol=BCHABC_USDT', 'limit': 157.81, 'stake_amount': 100, 'stake_currency': 'USDT', 'fiat_currency': 'USD'}
2018-12-20 05:38:38,174 - freqtrade.wallets - INFO - Wallets synced ...
2018-12-20 05:38:39,171 - freqtrade.wallets - INFO - Wallets synced ...
2018-12-20 05:38:43,063 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=7, pair=BCHABC/USDT, amount=0.63367340, open_rate=157.81000000, open_since=just now)
2018-12-20 05:38:44,072 - freqtrade.freqtradebot - WARNING - amount 0.6336700000000001 does not match amount 0.6336734047272036
2018-12-20 05:38:44,072 - freqtrade.freqtradebot - WARNING - could not update trade amount: Half bought? Amounts don't match
2018-12-20 05:38:44,075 - freqtrade.persistence - INFO - Updating trade (id=7) ...
2018-12-20 05:38:44,075 - freqtrade.persistence - INFO - LIMIT_BUY has been fulfilled for Trade(id=7, pair=BCHABC/USDT, amount=0.63367000, open_rate=157.81000000, open_since=just now).

2018-12-20 06:28:11,764 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market BCHABC/USDT.Tried to sell amount 0.63366 at rate 160.04 (total 101.4109464).Message: binance {"code":-2010,"msg":"Account has insufficient balance for requested action."}
2018-12-20 06:28:16,623 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market BCHABC/USDT.Tried to sell amount 0.63366 at rate 160.01 (total 101.3919366).Message: binance {"code":-2010,"msg":"Account has insufficient balance for requested action."}
2018-12-20 06:28:21,419 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market BCHABC/USDT.Tried to sell amount 0.63366 at rate 159.6 (total 101.132136).Message: binance {"code":-2010,"msg":"Account has insufficient balance for requested action."}
2018-12-20 06:28:26,431 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market BCHABC/USDT.Tried to sell amount 0.63366 at rate 159.78 (total 101.2461948).Message: binance {"code":-2010,"msg":"Account has insufficient balance for requested action."}
2018-12-20 06:28:31,436 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market BCHABC/USDT.Tried to sell amount 0.63366 at rate 159.67 (total 101.1764922).Message: binance {"code":-2010,"msg":"Account has insufficient balance for requested action."}
2018-12-20 06:28:36,441 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market BCHABC/USDT.Tried to sell amount 0.63366 at rate 159.
```

**Comment by asheetov** [comment]
The bad thing here is not wrong rounding or whatever, but fact that freqtrade report nothing to user about the issue. Missed stoploss is just silently ignoring, and user happily losing money without any notice through RPC.

**Comment by creslinux** [comment]
It's not a stop loss if it's not on the exchange, it's a sell order.

And best of hope code is working, price move is spotted in good time,
internet connection is up, exchange is not under ddos, there is some
liquidity remaining on the price step after all the exchange stop losses
are settled, the PC the bot is running on has not crashed

**Comment by mishaker** [comment]
@asheetov thanks for reporting. my take is that you are using trailing stoploss and when the bot tries to sell in bear market it is failing due to the slippage.
However setting sell order type to "market" should logically resolve the problem (see https://github.com/freqtrade/freqtrade/blob/develop/docs/configuration.md#understand-order_types)

My advice is turn the stoploss on exchange "True"  and also put sell to "market". 

As for your comment on bot should inform user, I cannot agree more. stoploss and trailing stoploss will undergo few refactoring but I am afraid it won't be before new year.

**Comment by asheetov** [comment]
I don't think the issue is related to slippage, because according to the logs, bot tried to sell the same (wrong) amount of currency several times at different prices. So it's not price-related issue, but amount related.

And, unfortunately, market order type doesn't help here #1428

**Comment by mishaker** [comment]
@asheetov please update your CCXT version. as you are behind the filter correction version. 
However I am not sure if it resolves the problem. 
Quick question: do you use base currency to pay the fees ?

**Comment by Wheaties466** [comment]
I believe I am having this same issue. I had to manually cancel an order(I'll open another issue as to why) On binance. Now it doesn't update my balance from binance and keeps spamming the error that I do not have enough funds.

WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market PHX/BTC.Tried to sell amount 8032.0 at rate 2.48e-06 (total 0.01991936).Message: binance {"code":-2010,"msg":"Account has insufficient balance for requested action."}

**Comment by xmatthias** [comment]
if you manually cancelled the order on the exchange you need to let the bot know and close / remove the trade in the database. 
The documentation has a cheatsheet [here](https://www.freqtrade.io/en/latest/sql_cheatsheet/) on how to do that.

**Comment by Wheaties466** [comment]
Got it I did that but figured since there isn't a built in function to do
so this want intended behavior.

Any ideas on why it would but and sell so many times in a row for a loss?

- Sent from my iPhone

**Comment by xmatthias** [comment]
bolt assumption without proof:
a) bad signals
b) rounding errors due to min-amount from the exchange (rounding up on buy - and because this is a very cheap coin that is > stoploss% - so it buys at that level and sells almost immediately (i assume in the next cycle)

I had this happen a few times with HOT - (which is since then on my blacklist ...)

**Comment by Wheaties466** [comment]
I blacklisted the coin and turned on the experimental feature to only sell
for a profit.

Is there additional logging I can enable to possibly see why this occurred.
It was a pretty Stark difference in price but I suppose it could be a
rounding error.

- Sent from my iPhone

**Comment by Wheaties466** [comment]
I just don't want to wake up to thousands of trades and no money left.

- Sent from my iPhone

**Comment by xmatthias** [comment]
The database (and telegram messages) should have the sell-reason (roi / stoploss / sell-signal) ... 
only sell for profit does not impact stoploss (just saying) ...

**Comment by Wheaties466** [comment]
9|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:15:22.106207|2019-01-30 14:15:31.296437||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
10|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:15:34.104425|2019-01-30 14:15:43.791615||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
11|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:15:46.632333|2019-01-30 14:15:55.819390||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
12|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:15:58.699380|2019-01-30 14:16:07.852004||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
13|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:16:10.714824|2019-01-30 14:16:20.416493||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
14|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:16:23.569230|2019-01-30 14:16:32.822851||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
15|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:16:35.683323|2019-01-30 14:16:45.232692||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
16|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.33400002|0.02|7991.281|2019-01-30 14:16:47.858248|2019-01-30 14:16:57.331063||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
17|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:17:00.199170|2019-01-30 14:17:09.378589||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
18|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:17:12.664875|2019-01-30 14:17:21.876437||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
19|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:17:24.733776|2019-01-30 14:17:33.998222||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
20|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:17:36.901630|2019-01-30 14:17:46.155915||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
21|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:17:49.034237|2019-01-30 14:17:58.219077||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
22|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:18:01.096536|2019-01-30 14:18:10.266107||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
23|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:18:13.128920|2019-01-30 14:19:05.184877||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
24|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:19:08.044202|2019-01-30 14:19:17.228284||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
25|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:19:20.086615|2019-01-30 14:19:29.273287||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5
26|binance|PHX/BTC|0|0.0|0.001|3.0e-06|2.5e-06|2.0e-06|2.49e-06|-0.334|0.02|7991.001|2019-01-30 14:19:32.166674|2019-01-30 14:19:41.436533||2.7e-06|2.7e-06|3.0e-06|stop_loss|DefaultStrategy|5


looks like stop loss. I just can't imagine why it would buy in the first place if the stop_loss is lower than the purchase price.

**Comment by xmatthias** [comment]
@Wheaties466 
I think it's rounding - @gaugau3000  explained it well n #1520 - this seems to happen with "very" low coins - which are a problem since exchanges have both a lot-size (minimum amount you are allowed to buy) as well as a max-precision.

there should probably be a check in place before buying - but to be clear, it would eliminate certain low-value coins completely.

##### for everyone else having this problem:
did you notice a log-entry like `Applying fee on amount for XXX failed: myTrade-Dict empty found` right after the buy for the coins in question?

**Comment by othaner** [comment]
I have the same issue. This is my log (precision filter deactivated):

```
2019-03-23 04:10:40,558 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=1, pair=BTT/BTC, amount=5050.00000000, open_rate=0.00000020, open_since=19 hours ago) ...
2019-03-23 04:10:40,559 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:40,592 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:40,593 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=BTT/BTC) buy=False sell=False
2019-03-23 04:10:40,594 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:40,724 - freqtrade.strategy.interface - DEBUG - Checking if trade is profitable..
2019-03-23 04:10:40,724 - freqtrade.freqtradebot - DEBUG - Found no sell signal for Trade(id=1, pair=BTT/BTC, amount=5050.00000000, open_rate=0.00000020, open_since=19 hours ago).
2019-03-23 04:10:40,724 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=2, pair=XLM/BTC, amount=38.00000000, open_rate=0.00002638, open_since=19 hours ago) ...
2019-03-23 04:10:40,725 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:40,762 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:40,764 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=XLM/BTC) buy=False sell=False
2019-03-23 04:10:40,764 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:41,228 - freqtrade.strategy.interface - DEBUG - using positive stop loss mode: 0.01 with offset 0.011 since we have profit 0.0264%
2019-03-23 04:10:41,229 - freqtrade.persistence - DEBUG - keeping current stop loss
2019-03-23 04:10:41,229 - freqtrade.persistence - DEBUG - XLM/BTC - current price 0.00002713, bought at 0.00002638 and calculatedstop loss is at: 0.00001319 initial stop at 0.00002713. trailing stop loss saved us: 0.00001394 and max observed rate was 0.00002740
2019-03-23 04:10:41,229 - freqtrade.strategy.interface - DEBUG - Required profit reached. Selling..
2019-03-23 04:10:41,735 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market XLM/BTC.Tried to sell amount 38.0 at rate 2.713e-05 (total 0.00103094).Message: binance Account has insufficient balancefor requested action.
2019-03-23 04:10:41,735 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=3, pair=FUEL/BTC, amount=345.00000000, open_rate=0.00000292, open_since=19 hours ago) ...
2019-03-23 04:10:41,737 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:41,786 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:41,787 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=FUEL/BTC) buy=False sell=False
2019-03-23 04:10:41,787 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:42,225 - freqtrade.strategy.interface - DEBUG - HIT STOP: current price at 0.000003, stop loss is 0.000003, initial stop loss was at 0.000001, trade opened at 0.000003
2019-03-23 04:10:42,225 - freqtrade.strategy.interface - DEBUG - trailing stop saved 0.000002
2019-03-23 04:10:42,225 - freqtrade.strategy.interface - DEBUG - Stop loss hit.
2019-03-23 04:10:42,735 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market FUEL/BTC.Tried to sell amount 345.0 at rate 2.91e-06 (total 0.00100395).Message: binance Account has insufficient balance for requested action.
2019-03-23 04:10:42,735 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=4, pair=WTC/BTC, amount=3.21000000, open_rate=0.00031400, open_since=18 hours ago) ...
2019-03-23 04:10:42,736 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:42,776 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:42,777 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=WTC/BTC) buy=False sell=False
2019-03-23 04:10:42,777 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:43,228 - freqtrade.strategy.interface - DEBUG - Checking if trade is profitable..
2019-03-23 04:10:43,228 - freqtrade.freqtradebot - DEBUG - Found no sell signal for Trade(id=4, pair=WTC/BTC, amount=3.21000000, open_rate=0.00031400, open_since=18 hours ago).
2019-03-23 04:10:43,228 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=5, pair=VET/BTC, amount=691.00000000, open_rate=0.00000146, open_since=18 hours ago) ...
2019-03-23 04:10:43,229 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:43,265 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:43,265 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=VET/BTC) buy=False sell=False
2019-03-23 04:10:43,266 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:43,960 - freqtrade.strategy.interface - DEBUG - using positive stop loss mode: 0.01 with offset 0.011 since we have profit 0.0185%
2019-03-23 04:10:43,960 - freqtrade.persistence - DEBUG - keeping current stop loss
2019-03-23 04:10:43,960 - freqtrade.persistence - DEBUG - VET/BTC - current price 0.00000149, bought at 0.00000146 and calculatedstop loss is at: 0.00000073 initial stop at 0.00000148. trailing stop loss saved us: 0.00000075 and max observed rate was 0.00000150
2019-03-23 04:10:43,961 - freqtrade.strategy.interface - DEBUG - Checking if trade is profitable..
2019-03-23 04:10:43,961 - freqtrade.freqtradebot - DEBUG - Found no sell signal for Trade(id=5, pair=VET/BTC, amount=691.00000000, open_rate=0.00000146, open_since=18 hours ago).
2019-03-23 04:10:43,962 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=6, pair=DOCK/BTC, amount=340.00000000, open_rate=0.00000297, open_since=18 hours ago) ...
2019-03-23 04:10:43,963 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:44,001 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:44,002 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=DOCK/BTC) buy=False sell=False
2019-03-23 04:10:44,002 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:44,226 - freqtrade.strategy.interface - DEBUG - HIT STOP: current price at 0.000003, stop loss is 0.000003, initial stop loss was at 0.000001, trade opened at 0.000003
2019-03-23 04:10:44,226 - freqtrade.strategy.interface - DEBUG - trailing stop saved 0.000002
2019-03-23 04:10:44,226 - freqtrade.strategy.interface - DEBUG - Stop loss hit.
2019-03-23 04:10:44,733 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Could not create limit sell order on market DOCK/BTC.Tried to sell amount 340.0 at rate 2.89e-06 (total 0.0009826).Message: binance {"code":-1013,"msg":"Filter failure: MIN_NOTIONAL"}
2019-03-23 04:10:44,734 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=7, pair=ZIL/BTC, amount=213.00000000, open_rate=0.00000475, open_since=18 hours ago) ...
2019-03-23 04:10:44,735 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:44,771 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:44,772 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=ZIL/BTC) buy=False sell=False
2019-03-23 04:10:44,772 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:45,229 - freqtrade.strategy.interface - DEBUG - HIT STOP: current price at 0.000005, stop loss is 0.000005, initial stop loss was at 0.000002, trade opened at 0.000005
2019-03-23 04:10:45,230 - freqtrade.strategy.interface - DEBUG - trailing stop saved 0.000002
2019-03-23 04:10:45,230 - freqtrade.strategy.interface - DEBUG - Stop loss hit.
2019-03-23 04:10:45,746 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market ZIL/BTC.Tried to sell amount 213.0 at rate 4.81e-06 (total 0.0010245299999999998).Message: binance Account has insufficient balance for requested action.
2019-03-23 04:10:45,746 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=8, pair=NULS/BTC, amount=6.00000000, open_rate=0.00016692, open_since=18 hours ago) ...
2019-03-23 04:10:45,747 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:45,785 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:45,786 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=NULS/BTC) buy=False sell=False
2019-03-23 04:10:45,786 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:46,226 - freqtrade.strategy.interface - DEBUG - HIT STOP: current price at 0.000173, stop loss is 0.000174, initial stop loss was at 0.000083, trade opened at 0.000167
2019-03-23 04:10:46,227 - freqtrade.strategy.interface - DEBUG - trailing stop saved 0.000091
2019-03-23 04:10:46,227 - freqtrade.strategy.interface - DEBUG - Stop loss hit.
2019-03-23 04:10:46,742 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market NULS/BTC.Tried to sell amount 6.0 at rate 0.0001733 (total 0.0010398).Message: binance Account has insufficient balance for requested action.
2019-03-23 04:10:46,743 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=9, pair=XVG/BTC, amount=587.00000000, open_rate=0.00000172, open_since=18 hours ago) ...
2019-03-23 04:10:46,744 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:46,783 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:46,784 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=XVG/BTC) buy=False sell=False
2019-03-23 04:10:46,784 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:47,225 - freqtrade.strategy.interface - DEBUG - Checking if trade is profitable..
2019-03-23 04:10:47,225 - freqtrade.freqtradebot - DEBUG - Found no sell signal for Trade(id=9, pair=XVG/BTC, amount=587.00000000, open_rate=0.00000172, open_since=18 hours ago).
2019-03-23 04:10:47,226 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=10, pair=EOS/BTC, amount=1.11000000, open_rate=0.00090480, open_since=18 hours ago) ...
2019-03-23 04:10:47,227 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:47,263 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:47,264 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=EOS/BTC) buy=False sell=False
2019-03-23 04:10:47,265 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:47,724 - freqtrade.strategy.interface - DEBUG - Checking if trade is profitable..
2019-03-23 04:10:47,725 - freqtrade.freqtradebot - DEBUG - Found no sell signal for Trade(id=10, pair=EOS/BTC, amount=1.11000000,open_rate=0.00090480, open_since=18 hours ago).
2019-03-23 04:10:47,725 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=11, pair=ADA/BTC, amount=71.00000000, open_rate=0.00001416, open_since=18 hours ago) ...
2019-03-23 04:10:47,726 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:47,768 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:47,769 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=ADA/BTC) buy=False sell=False
2019-03-23 04:10:47,769 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:48,228 - freqtrade.strategy.interface - DEBUG - using positive stop loss mode: 0.01 with offset 0.011 since we have profit 0.0720%
2019-03-23 04:10:48,228 - freqtrade.persistence - DEBUG - keeping current stop loss
2019-03-23 04:10:48,228 - freqtrade.persistence - DEBUG - ADA/BTC - current price 0.00001521, bought at 0.00001416 and calculatedstop loss is at: 0.00000708 initial stop at 0.00001519. trailing stop loss saved us: 0.00000811 and max observed rate was 0.00001534
2019-03-23 04:10:48,229 - freqtrade.strategy.interface - DEBUG - Required profit reached. Selling..
2019-03-23 04:10:48,736 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market ADA/BTC.Tried to sell amount 71.0 at rate 1.521e-05 (total 0.00107991).Message: binance Account has insufficient balancefor requested action.
2019-03-23 04:10:48,736 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=12, pair=LOOM/BTC, amount=53.00000000, open_rate=0.00001891, open_since=17 hours ago) ...
2019-03-23 04:10:48,737 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:48,780 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:48,781 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=LOOM/BTC) buy=False sell=False
2019-03-23 04:10:48,781 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:49,229 - freqtrade.strategy.interface - DEBUG - Checking if trade is profitable..
2019-03-23 04:10:49,229 - freqtrade.freqtradebot - DEBUG - Found no sell signal for Trade(id=12, pair=LOOM/BTC, amount=53.00000000, open_rate=0.00001891, open_since=17 hours ago).
2019-03-23 04:10:49,229 - freqtrade.freqtradebot - DEBUG - Handling Trade(id=13, pair=ZEC/BTC, amount=0.07200000, open_rate=0.01391100, open_since=17 hours ago) ...
2019-03-23 04:10:49,230 - freqtrade.strategy.interface - DEBUG - TA Analysis Launched
2019-03-23 04:10:49,264 - freqtrade.strategy.interface - DEBUG - Loop Analysis Launched
2019-03-23 04:10:49,265 - freqtrade.strategy.interface - DEBUG - trigger: 2019-03-23 04:05:00+00:00 (pair=ZEC/BTC) buy=False sell=False
2019-03-23 04:10:49,265 - freqtrade.freqtradebot - DEBUG - checking sell
2019-03-23 04:10:49,761 - freqtrade.strategy.interface - DEBUG - HIT STOP: current price at 0.013859, stop loss is 0.013990, initial stop loss was at 0.006955, trade opened at 0.013911
2019-03-23 04:10:49,762 - freqtrade.strategy.interface - DEBUG - trailing stop saved 0.007034
2019-03-23 04:10:49,762 - freqtrade.strategy.interface - DEBUG - Stop loss hit.
2019-03-23 04:10:50,233 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Could not create limit sell order on market ZEC/BTC.Tried to sell amount 0.072 at rate 0.013859 (total 0.000997848).Message: binance {"code":-1013,"msg":"Filter failure: MIN_NOTIONAL"}
```

**Comment by othaner** [comment]
I checked some of the unable-to-sell trades and found this:

``` 
2019-03-23 04:21:20,336 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market FUEL/BTC.Tried to sell amount 345.0 at rate 2.91e-06 (total 0.00100395).Message: binance Account has insufficient balance for requested action.
```
But in my Binance account I only have 344 FUEL.

Goes the same for XLM:
```
2019-03-23 04:21:19,339 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market XLM/BTC.Tried to sell amount 38.0 at rate 2.722e-05 (total 0.00103436).Message: binance Account has insufficient balancefor requested action.
```
Bot try to sell 38, while in my account I only have 37.

I don't know if this could help to debugging this.

**Comment by konqueror1** [comment]
> I checked some of the unable-to-sell trades and found this:
> 
> ```
> 2019-03-23 04:21:20,336 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market FUEL/BTC.Tried to sell amount 345.0 at rate 2.91e-06 (total 0.00100395).Message: binance Account has insufficient balance for requested action.
> ```
> 
> But in my Binance account I only have 344.65500000 FUEL.
> 
> Goes the same for XLM:
> 
> ```
> 2019-03-23 04:21:19,339 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create limit sell order on market XLM/BTC.Tried to sell amount 38.0 at rate 2.722e-05 (total 0.00103436).Message: binance Account has insufficient balancefor requested action.
> ```
> 
> Bot try to sell 38, while in my account I only have 37.96200000.
> 
> Same for ADA tried to sell amount 71.0, only have 70.92900000.
> I don't know if this could help to debugging this.

I experience the same problem. I think it is because the trading fee is not deducted from amount (37.962 is 99.9% of 38.0 and 0.1% is standart Binance fee).
I am not sure it happens only when using bnb for fees and there is not enough bnb in balance so the fee is taken from asset. May be we have to check if amounts in exchange and in the sql are the same and update sql if needed.

**Comment by xmatthias** [comment]
this check is in place here:
https://github.com/freqtrade/freqtrade/blob/ba558b2d75ce631c9ae02af05386c85f52e35ab4/freqtrade/freqtradebot.py#L573

apparently, it's not working as it should.
I struggle to recreate this as I do have BNB in my account.

The problem is that you'll need the logs from the buy to the corresponding trade, which no one with this problem seems to have been able to capture so far.

**Comment by konqueror1** [comment]
This is what I have from my screen session but I think you will need debug log, that I don't have and don't know how to turn on.
> 2019-03-30 05:34:06,760 - freqtrade.freqtradebot - INFO - Buy signal found: about create a new trade with stake_amount: 0.025 ...
2019-03-30 05:34:06,760 - freqtrade.freqtradebot - INFO - Using Last Ask / Last Price
2019-03-30 05:34:07,550 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': buy, 'exchange': 'Binance', 'pair': 'DGD/BTC', 'limit': 0.005019,
'stake_amount': 0.02499963, 'stake_currency': 'BTC', 'fiat_currency': 'USD'}
2019-03-30 05:34:08,045 - freqtrade.wallets - INFO - Wallets synced.
2019-03-30 05:34:10,540 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Insufficient funds to create stop_loss_limit sell order on market DGD/BTC.
Tried to sell amount 4.981 at rate 0.004721 (total 0.023515301000000002).Message: binance Account has insufficient balance for requested action.

And form Binance:
> 

Date | Pair | Type | Price | Filled | Fee | Total
-- | -- | -- | -- | -- | -- | --
2019-03-30 05:34:07 | DGD/BTC | Buy | 0.005019 | 4.981 | 0.00498100 DGD | 0.02499963 BTC

**Comment by xmatthias** [comment]
@konqueror1 
the logs indicate that it's not a sell-command, but a stoploss-order.

This could be a problem in the first iteration (if the trade is not filled completely yet ...).
Did the bot succeed to place the stoplooss-order a little later (a few iterations / seconds)?

**Comment by creslinux** [comment]
These can be common:
 - Insufficient balance when the sell would leave no funds for fees. Its
returned as an Exception, so loop on this and reset at amount * 0.997
 - Cannot set, as would immediately be triggered. When there is a large
spread between ASK and BID, or price has fallen. Again us the same
exception to walk down price in 0.2% decrements until have a stop.
I tackled the same in the below structure, not had a stop miss since.
A Stop that cover 99.7% the position and/or a fraction % off the ideal is
far better than no stop

```
For ... prices / amounts :
  log my_try price / amount ..
   Try
        _ret =.....order stuff
        log got it
        Break
   Except e
        log e - get the detail
        Continue
Else:
    log warning all prices amounts failed
    Return None

my_order =  _ret. # success do stuff
```

**Comment by creslinux** [comment]
This is how the exception is returned.

**Comment by konqueror1** [comment]
@xmatthias 
the bot succeeded when I corrected the amount in the DB. The available amount in DB was wrong from the beginning.
I don't know the logic for getting info like amounts, fees etc but I believe it does not cover all the cases. I will try to debug what get_real_amount actually does.

**Comment by mishaker** [comment]
I didn’t follow this thread in details but I am wondering why we don’t just
update the amount in DB right after buy is executed?

**Comment by asheetov** [comment]
Just a suggestion of different approach for selling algorithm, which works perfect for me in Zenbot (honestly I give up with freqtrade after losing money because of this issue).

Don't make the bot to calculate "exactly right" amount of selling position considering fees, rounding and whatever else. Just sell "all of them". For flexibility you can add the new config option like "sell percent of open position" and place sell orders according that percent of funds in account.

**Comment by creslinux** [comment]
its fees, place the stop-limit order it in a loop that continues on
exception to handle this. 1
1st - try to sell 100% - if user has BNB and is fee paying with BNB this
will work.
2nd -  try to sell 99.7%the position amount - if user has no BNB or is not
paying with BNB, the amount it their full balance, this will work.

This is true to stops and sells.

Next, in stops alone:
User will also hit stop would trigger immediately exception on large spot
spread or price is falling fast.
Use the same loop to lower the  stop-trigger and limit price here too.

Its a mistake to think the bot can calculate this upfront always, the
market is dynamic and does not care.
The exceptions should be caught and handled appropriately - its easier to
ask for forgiveness in the code paradigm

Its important a position is protected with stop and a bad position exits
than leaving money with no stop at all.
Risk capital is 100% without a stop.

Any dust from selling just shy of full amount can be swept up later by
user.

**Comment by xmatthias** [comment]
it's not that we try to calculate the Amount, we have code in place which detects the real bought amount based on the buy order id (see my previous comment).

However, due to a bug, that code was never executed for buys which filled immediately (see #1720 for details), which caused the relevant code to be never executed during a buy - so the trade-amount was never updated.

I did multiple buys/sells with the above PR, all with BNB disabled (and the amount did update to the correct value).

Side note: Retrying with lower amounts until it works is a brute-force method, without really caring about the root problem.

**Comment by creslinux** [comment]
The root of the problem is returned exceptions are not being handled, but
ignored. The impact of ignoring exceptions being sent is users real money
is left at risk for no reason.

The use case for triggering the scenario is.
I have 10LTC total and my stop is to sell 10LTC or my sell is to sell
10LTC. My fees are to be paid in LTC.  I can't sell 10 as I need to pay the
fee too. Binance returns a insufficient funds error.

Lowering the amount by .3% covers the fee, when the exception is returned
protects the capital.

The dust remaining can be swept up after.

Really when the exception returned states insignificant funds, lowering the
amount is the correct and obvious solution. It's a million times better
than not protecting money. It's not brute Force either, because it'sa
targetted handling, which works.

**Comment by creslinux** [comment]
Another factor to be aware of is to never set a buy for less than
min_notional plus your stop.

If min notional is 5 and you buy 5 you can't place a stop where by if the
limit price is met it only returns 4.8 , as example. Stop liquidation price
must also meet min notional.

**Comment by xmatthias** [comment]
we're not ignoring the exception, but stupidly recreating the order with another amount *IS NOT THE SOLUTION*.

Investigating the problem, finding out WHY the amounts don't match, that is the challange. 
Stupidly retrying until it works is something a schoolgirl can do, i don't need a bot for that.

**Comment by creslinux** [comment]
I've explained exactly what the scenario is.

And at some point it may be hit. Regardless your opinion of school
girls???  the exchange is reality, and it it says no there are insufficient
funds in position not protected.

If you want to check all balances first, be totally certain nothing else
can possibly occur in parallel to upset that, make a targeted stop either
for 100% or 99.7% .. its a way.
It will be slower, take more calls, still leave the possibility of the
exception occuring. Needing to be handled.

**Comment by creslinux** [comment]
https://en.m.wikipedia.org/wiki/Ada_Lovelace

**Comment by creslinux** [comment]
Here is an example of the error being handled.
Reduced from 186.5 100% amount, as no money for fees, placed at 185.9
Time took was a ms, then the position was protected.
Any solution that leaves the position unprotected or unprotected for longer
isn't better

Its a solution, it's a benchmark - happy for it to be beaten.

**Comment by mishaker** [comment]
@creslinux 
> The use case for triggering the scenario is.
I have 10LTC total and my stop is to sell 10LTC or my sell is to sell
10LTC. My fees are to be paid in LTC.  I can't sell 10 as I need to pay the
fee too. Binance returns a insufficient funds error.

This is simply a misdiagnosis. 
Fees are not extracted beforehand from Binance. (i.e. the amount is not rejected because it is more than the (amount - fees). from Binance docs, a FULL response to a sell order: 

```JSON
{
  "symbol": "BTCUSDT",
  "orderId": 28,
  "clientOrderId": "6gCrw2kRUAF9CvJDGP16IP",
  "transactTime": 1507725176595,
  "price": "1.00000000",
  "origQty": "10.00000000",
  "executedQty": "10.00000000",
  "cummulativeQuoteQty": "10.00000000",
  "status": "FILLED",
  "timeInForce": "GTC",
  "type": "MARKET",
  "side": "SELL",
  "fills": [
    {
      "price": "4000.00000000",
      "qty": "1.00000000",
      "commission": "4.00000000",
      "commissionAsset": "USDT"
    },
    {
      "price": "3999.00000000",
      "qty": "5.00000000",
      "commission": "19.99500000",
      "commissionAsset": "USDT"
    },
    {
      "price": "3998.00000000",
      "qty": "2.00000000",
      "commission": "7.99600000",
      "commissionAsset": "USDT"
    },
    {
      "price": "3997.00000000",
      "qty": "1.00000000",
      "commission": "3.99700000",
      "commissionAsset": "USDT"
    },
    {
      "price": "3995.00000000",
      "qty": "1.00000000",
      "commission": "3.99500000",
      "commissionAsset": "USDT"
    }
  ]
}
```

Note the origQty, executedQty (10 BTC) and the sum of all trades (1 +5 +2 +1 +1). 
The safest and the most pertinent approach is to update the trade amount immediately if the fee is paid by the base currency. coded here: 

```python
if trade.pair.startswith(order['fee']['currency']): # <-- checks if the fee currency is the base currency
                new_amount = order_amount - order['fee']['cost']
                logger.info("Applying fee on amount for %s (from %s to %s) from Order",
                            trade, order['amount'], new_amount)
                return new_amount
```

This is what was skipped before @xmatthias found the bug today and corrected it. 

> The root of the problem is returned exceptions are not being handled, but
ignored. The impact of ignoring exceptions being sent is users real money
is left at risk for no reason.

Your comment is completely nonsense technically and condescending socially. start with the technical part: no exception is ignored or unhandled in this project, it is just a matter of reporting it and finding the right solution rather than coding a workaround which causes more problems. As for the social part: finally someone found the bug hanging since last November and corrected it. you didn't even take time to read it thoroughly before commenting "Me Myself and I ..." and unfortunately this is not the first time you show this attitude.

**Comment by creslinux** [comment]
@misagh
Take a moment to read the output you've posted.
And then what I posted and try again.

The condition is when it is the LAST 10, in your example.
if your account has ONLY 10 - in total not 1 cent more.
Try and set your sell or stop for all 10. And watch binance tell you in an
`exception` insufficient funds

which is no being handled - its not condescending  - unlike the school girl
sexist comment .... to point out the exception is not being handled.
hopefully the school girl would have a stop on her funds and not handing
others money out to dry.

**Comment by creslinux** [comment]
The word you're missing from my statement is "total", can replace this with
"only" if it makes it clearer.

Use case  in more detail. This should not be that hard to understand.
 "account" in TOTAL (or ONLY) has $10 value
** you can have more but if its tied into other currencies/positions this
is the same, it cannot be spent on fees for this position.
Try and spend or set a stop for $10.

**Comment by creslinux** [comment]
The solution being put forward by xmats is wrong btw - if what @misagh
posted inline was it?
Its moved the problem from handling error state bad to handling good state
poorly.

Use-case 1
If i have BNB or USDT free to pay the fees when I exit 10LTC/USDT(as
example) i want to exit all 10 LTC. Not exit 10LTC - fees.
This would leave me with LTC dust. Because the fees would be paid by the
USDT or BNB free in my account already.

Use-case 2
I have not BNB or USDT to pay the fees when i exit 10LTC/USDT(as example) i
want to exit all 10LTC.
Now the amount is reduced, if i read it correctly, minus fees. This is
probably good behaviour

The problem is by Always subtracting fees from sell amount, is when the
user HAS free to pay fees this is not desired, it leaves a fees worth of
base asset behind in every trade. Dust.

**Comment by xmatthias** [comment]
>  unlike the school girl sexist comment 

You are right @creslinux, edited that comment now to say "schoolgirl/schoolboy". Wasn't intendet to be sexist, sorry.

@creslinux again, **please edit your comments instead of posting 3 comments in quick succession** if you need to make corrections or additions right after posting. This creates 3 alerts for everyone subscribed to this issue, where on ewould be enough. 
It's fine to recomment on your own answer if that was more than a day ago.

> he solution being put forward by xmats is wrong btw

Please take the time to understand the solution before you post misleading comments.
My solution is not new, but was there and working before we added different time_in_force methods, which introduced the bug causing this.
The solution gathers the trades connected to the buy-order, analyzes if the fees have been payed in stake-currency or not (payed in BNB/USDT) - and then adjusts the amount eventually.

The message/ return value from @mishaker is not what is used for this calculation.
We use the `"fees"` element of order if it's there (hint: it's not for binance) or get the trades for this order, and use the `"fee"` element in this return value.

The result of the "fetch_my_trades" is something like the following:
It includes the "fees" dict, and if the currency is MTL - we subtract the amount.

## Dict in case of BNB fees

``` json
[{'info': {'symbol': 'MTLETH',
   'id': 1234,
   'orderId': 1234,
   'price': '0.00527100',
   'qty': '18.96000000',
   'quoteQty': '0.09993816',
   'commission': '0.00060196',
   'commissionAsset': 'BNB',
   'time': 1554032217497,
   'isBuyer': True,
   'isMaker': False,
   'isBestMatch': True},
  'timestamp': 1554032217497,
  'datetime': '2019-03-31T11:36:57.497Z',
  'symbol': 'MTL/ETH',
  'id': '1234',
  'order': '1234',
  'type': None,
  'takerOrMaker': 'taker',
  'side': 'buy',
  'price': 0.005271,
  'amount': 18.96,
  'cost': 0.09993816,
  'fee': {'cost': 0.00060196, 'currency': 'BNB'}}]

```

### dict in  case of non-BNB fees:
```
[{'info': {'symbol': 'MTLETH',
   'id': 5543,
   'orderId': 5555,
   'price': '0.00428600',
   'qty': '23.33000000',
   'quoteQty': '0.09999238',
   'commission': '0.02333000',
   'commissionAsset': 'MTL',
   'time': 1554053985003,
   'isBuyer': True,
   'isMaker': False,
   'isBestMatch': True},
  'timestamp': 1554053985003,
  'datetime': '2019-03-31T17:39:45.003Z',
  'symbol': 'MTL/ETH',
  'id': '5543',
  'order': '5555',
  'type': None,
  'takerOrMaker': 'taker',
  'side': 'buy',
  'price': 0.004286,
  'amount': 23.33,
  'cost': 0.09999237999999999,
  'fee': {'cost': 0.02333, 'currency': 'MTL'}},
]
```

Note: There can be more than one trade per side, in which case the amount is aggregated - but there have not been in this case and for simplicity of .

**Comment by creslinux** [comment]
This isn't going to fix the problem then.

Looking at fees for when bought into a position is not going help when
trying to sell 100% of a position when there is no other funds available to
pay the fee.

For absolute simplicity, if the account has only 1 balance. All other
wallets are zero.

10 abc/ETH

The fees are to be paid in abc's

Trying to set a stop for 10abc will fail with insufficient balance.

The error is not as simple as trying to sell more than is in the free
balance.
It is trying to sell all that is in the free balance and leave none to pay
the fees.

**Comment by xmatthias** [comment]
From your answer it's not 100% clear, but i assume you only have 10abc in your account.

Your sell-order for 10abc will succeed, since you are paying fees in the amount you receive - which is ETH in your example. So you'll not get 10abs's worth of ETH - but 10-<fee>abc's worth of ETH.

For binance, fees are paid in the currency you're about to receive (or BNB).

## another calculation sample, hopefully clearer
Let's assume i have BNB disabled for this post, and for simplicity, the rate between XRP and ETH is 1.

If i buy 100 XRP, then i will receive 99.75 XRP, and have paid 0.25 XRP in feeds.
The bot detects this - and sets the amount of XRP i have to 99.75.

When trying to sell - or creating a stoploss order - I will place my order with 99.75 XRP.
I will however not receive 99.75 ETH, but 99.49 ETH, since fees are subtracted by from this.

I do not have to do this calculation myself and place the sell-order with 99.49, which is what you're suggesting, but the exchange subtracts this from the currency i am about to receive (both ways, when buying or selling).

Note: this is a fictional example, XRP ETH rate is not 1 - but it simplifies the calculation.

**Comment by creslinux** [comment]
Everything you're saying makes sense.
The challenge I have with this is the code im using does not track, tot up,
balances, ergo it cant be wrong.
We receive the balances from Binance, after each trade, a full balance is
received this overwrites total/free/used it full - no adding / subtracting
is involved.
As an aside the update always includes all other asset balances too,
binance send the full set of wallets in 1 payload.

Selling or Stop on 100% `free` of an asset works almost every time, 29
trades in 30 as example -  apart from when there is no other funds for
fees, if it's the last of account balance being committed, then i get the
error. Is it True that fees are Always taken from received asset?

**Comment by mishaker** [comment]
yep, that is what we are trying to explain here :) fees are always taken from received amount. your mistake was you thought you have to calculate the fee beforehand cause binance it reduces from your base amount on market price of base/BNB. as you can never be sure of this rate then you came to conclusion that a loop should be baked. your hypothesis was wrong. sell amount is always fixed, fees are reduced from received asset.

**Comment by creslinux** [comment]
Misagh

Thats missing a point - the error IS received and ONLY when its the last of
balance in a position.
I cannot be trying to sell more than my free - as binance tells me what it
is - i dont calculate /hold it as FT is.

I can posts literally a hundred examples as its all logged where there is
free other balance and 100% position stops work, are set
I can also post the same where the error is received when is the last of
balance.

**Comment by mishaker** [comment]
@iuvbio this one. I think it is the root of your problem on #1723

**Comment by mishaker** [comment]
I am going to close this issue as I am sure it is solved by #1720 
Please do not hesitate to open it if any of you think differently.

**Comment by iuvbio** [comment]
Mmm, I'm not sure, I did not set FOK for time_in_force..

**Comment by xmatthias** [comment]
It's not an issue with FOK - but that this applied for every trade which was immediately closed, no matter if it's IOC, FOK, or GTC (the default).
